calculator.rb

```diff
@@ -0,0 +1,174 @@
+	else
+		puts "I do not recognize the operation \"#{operation}\". Please enter: addition, subtraction, multiplication, division, exponent, or modulus."
+		operation = gets.chomp.downcase
+	end
+end
+
+# Ask the user for a number and store the input in a variable.
+puts "Enter your first number:"
+first_number_raw = gets.chomp
+
+# Reset the boolean variable to store the validity of the number
+valid = false
+
+# Run loop to check for number validity
+until valid == true
+	# Check to see if the first and last characters are (), then run evaluate_parens and feed the result into first number
+	if first_number_raw[0] == "(" && first_number_raw[-1] == ")"
+		first_number = evaluate_parens(first_number_raw).to_s
+		valid = true
+	# Check to make sure the input is a number
+	# If the input, when converted to a float then back to a string, or to an integer, then back to a string, does not retrain its identity
+	# or if the input is an empty string, then the number is invalid and a request is made for more input.
+	# Delete commas before processing to allow for numbers inputted with commas, like "2,000"
+	elsif first_number_raw.delete(",").to_f.to_s != first_number_raw.delete(",") \
+		&& first_number_raw.delete(",").to_i.to_s != first_number_raw.delete(",")
+		puts "Please enter a valid number."
+		first_number_raw = gets.chomp
+	# Otherwise the input is a valid number and the loop ends
+	else
+	 	first_number = first_number_raw
+		valid = true
+	end
+end
+
+# Perform the same operations to check for validity with the second number inputted.
+puts "Enter your second number:"
+second_number_raw = gets.chomp
+
+# Reset the validity variable
+valid = false
+
+# Run loop to check for number validity
+while valid == false
+	if second_number_raw[0] == "(" && second_number_raw[-1] == ")"
+		second_number = evaluate_parens(second_number_raw).to_s
+		valid = true
+	elsif second_number_raw.delete(",").to_f.to_s != second_number_raw.delete(",") \
+		&& second_number_raw.delete(",").to_i.to_s != second_number_raw.delete(",")
+		puts "Please enter a valid number."
+		second_number_raw = gets.chomp
+	else
+	 	second_number = second_number_raw
+		valid = true
+	end
+end
+
+# run the calculate operation on the data collected and processed
+main_output, main_operator = calculate(operation, first_number, second_number)
+
+# Check if operation is exponent AND if the second number is an integer
+# If so, create full printout of operation (eg. 2 ^ 3 = 2 * 2 * 2)
+if $supported_operations[:exponent].include?(operation) && second_number.include?(".") == false
+	printout = "#{first_number}"
+	for i in 1...second_number.to_i
+		printout += " * #{first_number}"
+	end
+# And print the entire printed-out calculation
+	puts "#{first_number_raw} ^ #{second_number_raw} = #{printout} = #{main_output}"
+# For all other operations, print the calculation and output.
+else
+	puts "#{first_number_raw} #{main_operator} #{second_number_raw} = #{main_output}"
```

Comment:
I like having a single puts statement -- exponent isn't more important than the other variables. So I could generate expand_exponent if the operator was exponent, and it would only print if it existed? Nice, and cleaner than how I did it.